middle_end/flambda/inlining/inlining_report.ml

```diff
@@ -0,0 +1,176 @@
+      Misc.fatal_errorf "incoherent depth at end of inlining report"
+
+  (* Entering a function declaration (possibly nested) *)
+  | { dbg; decision = At_function_declaration {
+      pass = Before_simplify; code_id; decision; } } :: r ->
+    Format.fprintf fmt "%a Definition of %s{%a}@\n"
+      stars depth (Code_id.name code_id) Debuginfo.print dbg;
+    Format.fprintf fmt "%a @[<v>Before simplification:@ @ %a@]@\n@\n"
+      stars (depth + 1)
+      Inlining_decision.Function_declaration_decision.report decision;
+    print ~depth:(depth + 1) fmt r
+
+  (* Function call *)
+  | { decision = At_call_site C_call; dbg; } :: r ->
+    Format.fprintf fmt "%a @[<v>%s of %s{%a}@ @ %s@ %s@]@\n@\n"
+      stars depth
+      (if depth = 0 then "Toplevel application" else "Application")
+      "<unknown function>" Debuginfo.print dbg
+      "The C function call has not been inlined"
+      "because the optimizer cannot inline C calls";
+    print ~depth fmt r
+  | { decision = At_call_site Method; dbg; } :: r ->
+    Format.fprintf fmt "%a @[<v>%s of %s{%a}@ @ %s@ %s@]@\n@\n"
+      stars depth
+      (if depth = 0 then "Toplevel application" else "Application")
+      "<unknown function>" Debuginfo.print dbg
+      "The method call has not been inlined"
+      "because the optimizer cannot inline method calls";
+    print ~depth fmt r
+  | { decision = At_call_site Unknown_function; dbg; } :: r ->
+    Format.fprintf fmt "%a @[<v>%s of %s{%a}@ @ %s@ %s@]@\n@\n"
+      stars depth
+      (if depth = 0 then "Toplevel application" else "Application")
+      "<unknown function>" Debuginfo.print dbg
+      "The function call has not been inlined"
+      "because the optimizer had not enough information about the function";
+    print ~depth fmt r
+  | { decision = At_call_site (Non_inlinable_function { code_id; });
+      dbg; } :: r ->
+    Format.fprintf fmt "%a @[<v>%s of %s{%a}@ @ %s@ %s@]@\n@\n"
+      stars depth
+      (if depth = 0 then "Toplevel application" else "Application")
+      (Code_id.name code_id) Debuginfo.print dbg
+      "The function call has not been inlined"
+      "because its definition was deemed not inlinable";
+    print ~depth fmt r
+  | { decision = At_call_site (Inlinable_function { code_id; decision; });
+      dbg; } :: r ->
+    Format.fprintf fmt "%a @[<v>%s of %s{%a}@ @ %a@]@\n@\n"
+      stars depth
+      (if depth = 0 then "Toplevel application" else "Application")
+      (Code_id.name code_id) Debuginfo.print dbg
+      Inlining_decision.Call_site_decision.report decision;
+    print ~depth fmt r
+
+  (* Exiting a function_declaration (possibly nested) *)
+  | { dbg ; decision = At_function_declaration {
+      pass = After_simplify; code_id; decision; } } :: r ->
+    Format.fprintf fmt "%a @[<v>After simplification of %s{%a}:@ @ %a@]@\n@\n@\n"
+      stars depth (Code_id.name code_id) Debuginfo.print dbg
+      Inlining_decision.Function_declaration_decision.report decision;
+    print ~depth:(depth - 1) fmt r
```

Comment:
I would prefer if this case was close to the other `At_function_declaration` case.

Reply:
code moved